advisor/src/main/kotlin/advisors/VulnerableCode.kt

```diff
@@ -0,0 +1,168 @@
+                            AdvisorSummary(startTime, endTime)
+                        )
+                    )
+                }
+            }.toMap()
+        } catch (e: Exception) {
+            e.showStackTrace()
+
+            val now = Instant.now()
+            val failedResults = listOf(
+                AdvisorResult(
+                    vulnerabilities = emptyList(),
+                    advisor = AdvisorDetails(advisorName),
+                    summary = AdvisorSummary(
+                        startTime = now,
+                        endTime = now,
+                        issues = listOf(
+                            createAndLogIssue(
+                                source = advisorName,
+                                message = "Failed to retrieve security vulnerabilities from $advisorName: " +
+                                        e.collectMessagesAsString()
+                            )
+                        )
+                    )
+                )
+            )
+            packages.associateWith { failedResults }
+        }
+    }
+
+    /**
+     * Construct a map from package URLs to the vulnerabilities of this package based on a
+     * [map with vulnerability references][packagesToVulnerabilityRefs] using the specified [service]. This
+     * function looks up the details of the vulnerabilities the references point to.
+     */
+    private suspend fun fetchVulnerabilityDetails(
+        packagesToVulnerabilityRefs: MutableMap<String, VulnerableCodeService.Vulnerabilities>,
+        service: VulnerableCodeService
+    ): Map<String, List<Vulnerability>> {
+        val vulnerabilityIds = packagesToVulnerabilityRefs.values.flatMap { it.unresolvedVulnerabilities }
+            .map { it.url }
+        val vulnerabilityDetails = resolveVulnerabilities(service, vulnerabilityIds)
+        return packagesToVulnerabilityRefs
+            .filterNot { it.value.unresolvedVulnerabilities.isEmpty() }
+            .map { entry ->
+                val vulnerabilities = entry.value.unresolvedVulnerabilities.mapNotNull { vulnerabilityDetails[it.url] }
+                entry.key to vulnerabilities
+            }.toMap()
+    }
+
+    /**
+     * Resolve the given [vulnerabilityUrls] by querying the corresponding vulnerability IDs from the VulnerableCode
+     * [service]. Return a map with the URLs keys and the resolved vulnerability details as values.
+     */
+    private suspend fun resolveVulnerabilities(
+        service: VulnerableCodeService,
+        vulnerabilityUrls: List<String>
+    ): Map<String, Vulnerability> =
+        coroutineScope {
+            vulnerabilityUrls.map { extractVulnerabilityId(it) }
+                .map { async { service.getVulnerability(it) } }
+                .awaitAll()
+                .map { v ->
+                    Vulnerability(id = v.cveId.orEmpty(), url = URI(v.url), severity = v.cvss ?: SEVERITY_UNDEFINED)
+                }.associateBy { "${it.url}" }
```

Comment:
Nit, this should be more compact:

                .map { Vulnerability(it.cveId.orEmpty(), it.cvss ?: SEVERITY_UNDEFINED, URI(it.url)) }
                .associateBy { "${it.url}" }